dcpquery/db/__init__.py

```diff
@@ -174,8 +136,126 @@ def create_recursive_process_functions_in_db():
+                                        FROM bundle_file_links
+                                        WHERE bundle_fqid = NEW.bundle_fqid
+                                        AND file_fqid = NEW.file_fqid
+                                    )
+                                    DO INSTEAD NOTHING;
+                                """
+    upsert_process_rule_sql = """
+                                    CREATE OR REPLACE RULE process_table_ignore_duplicate_inserts AS
+                                        ON INSERT TO processes
+                                            WHERE EXISTS (
+                                              SELECT 1
+                                            FROM processes
+                                            WHERE process_uuid = NEW.process_uuid
+                                        )
+                                        DO INSTEAD NOTHING;
+                                    """
+
+    get_all_children_function_sql = """
+                        CREATE or REPLACE FUNCTION get_all_children(IN parent_process_uuid UUID)
+                            RETURNS TABLE(child_process UUID) as $$
+                              WITH RECURSIVE recursive_table AS (
+                                SELECT child_process_uuid FROM process_join_table
+                                WHERE parent_process_uuid=$1
+                                UNION
+                                SELECT process_join_table.child_process_uuid FROM process_join_table
+                                INNER JOIN recursive_table
+                                ON process_join_table.parent_process_uuid = recursive_table.child_process_uuid)
+                            SELECT * from recursive_table;
+                            $$ LANGUAGE SQL;
+                                    """
+    get_all_parents_function_sql = """
+                        CREATE or REPLACE FUNCTION get_all_parents(IN child_process_uuid UUID)
+                            RETURNS TABLE(parent_process UUID) as $$
+                              WITH RECURSIVE recursive_table AS (
+                                SELECT parent_process_uuid FROM process_join_table
+                                WHERE child_process_uuid=$1
+                                UNION
+                                SELECT process_join_table.parent_process_uuid FROM process_join_table
+                                INNER JOIN recursive_table
+                                ON process_join_table.child_process_uuid = recursive_table.parent_process_uuid)
+                            SELECT * from recursive_table;
+                            $$ LANGUAGE SQL;
+"""
+
+    def create_upsert_rules_in_db(cls):
+        config.db_session.execute(
+            cls.upsert_bundle_file_link_rule_sql + cls.upsert_bundle_rule_sql + cls.upsert_file_rule_sql
+        )
+        # TODO do I need to create an upsert rule for bundle_file_links or process_join_table?
```

Comment:
Can this TODO be resolved? If not, can you rephrase it like this:
```
# TODO (MDunitz) need to create an upsert rule for bundle_file_links or process_join_table?
```

Reply:
👍